File: src/test/java/leets/leenk/domain/feed/domain/service/FeedUpdateServiceTest.java
package leets.leenk.domain.feed.domain.service;

import leets.leenk.domain.feed.application.dto.request.FeedUpdateRequest;
import leets.leenk.domain.feed.domain.entity.Feed;
import leets.leenk.domain.feed.domain.entity.Reaction;
import leets.leenk.domain.feed.test.FeedTestFixture;
import leets.leenk.domain.feed.test.ReactionTestFixture;
import leets.leenk.domain.feed.test.UserTestFixture;
import leets.leenk.domain.user.domain.entity.User;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;

import static org.assertj.core.api.Assertions.assertThat;

public class FeedUpdateServiceTest {
    private final FeedUpdateService feedUpdateService = new FeedUpdateService();

    @Test
    @DisplayName("feedUpdate 테스트")
    void feedUpdate1() {
        // given
        User user = UserTestFixture.createUser(1L, "me");
        Feed feed = FeedTestFixture.createFeed(1L, user);

        FeedUpdateRequest request = new FeedUpdateRequest("\n\nhello\n\n\nworld", null, null);

        // when
        feedUpdateService.update(feed, request);

        // then
        assertThat(feed.getDescription()).isEqualTo("hello\nworld");
    }

    @Test
    @DisplayName("feedUpdate 테스트 - description이 null일 때 유지")
    void feedUpdate2() {
        // given
        User user = UserTestFixture.createUser(1L, "me");
        Feed feed = FeedTestFixture.createFeed(1L, user);
        feed.updateDescription("old");

        FeedUpdateRequest request = new FeedUpdateRequest(null, null, null);

        // when
        feedUpdateService.update(feed, request);

        // then
        assertThat(feed.getDescription()).isEqualTo("old");
    }

    @Test
    @DisplayName("updateTotalReaction 테스트")
    void updateTotalReaction() {
        // given
        User user = UserTestFixture.createUser(1L, "me");
        Feed feed = FeedTestFixture.createFeed(1L, user);
        Reaction reaction = ReactionTestFixture.createReaction(feed, user, 2L);

        // when
        feedUpdateService.updateTotalReaction(feed, reaction, user, 3L);

        // then
        assertThat(feed.getTotalReactionCount()).isEqualTo(3L);
        assertThat(reaction.getReactionCount()).isEqualTo(5L);
        assertThat(user.getTotalReactionCount()).isEqualTo(3L);
    }
}

File: src/test/kotlin/leets/leenk/domain/feed/domain/service/FeedUpdateServiceTest.kt
package leets.leenk.domain.feed.domain.service

import io.kotest.core.spec.style.BehaviorSpec
import io.kotest.matchers.shouldBe
import leets.leenk.domain.feed.application.dto.request.FeedUpdateRequest
import leets.leenk.domain.feed.test.FeedTestFixture
import leets.leenk.domain.feed.test.ReactionTestFixture
import leets.leenk.domain.feed.test.UserTestFixture

class FeedUpdateServiceTest :
    BehaviorSpec({
        val feedUpdateService = FeedUpdateService()

        Given("피드와 업데이트 요청이 주어졌을 때") {
            val user = UserTestFixture.createUser(1L, "me")
            val feed = FeedTestFixture.createFeed(1L, user)
            val request = FeedUpdateRequest("\n\nhello\n\n\nworld", null, null)

            When("피드를 업데이트하면") {
                feedUpdateService.update(feed, request)

                Then("description이 정리되어 업데이트되어야 한다") {
                    feed.description shouldBe "hello\nworld"
                }
            }
        }

        Given("기존 description이 있는 피드와 null description 요청이 주어졌을 때") {
            val user = UserTestFixture.createUser(1L, "me")
            val feed = FeedTestFixture.createFeed(1L, user)
            feed.updateDescription("old")
            val request = FeedUpdateRequest(null, null, null)

            When("피드를 업데이트하면") {
                feedUpdateService.update(feed, request)

                Then("기존 description이 유지되어야 한다") {
                    feed.description shouldBe "old"
                }
            }
        }

        Given("피드, 리액션, 사용자가 주어졌을 때") {
            val user = UserTestFixture.createUser(1L, "me")
            val feed = FeedTestFixture.createFeed(1L, user)
            val reaction = ReactionTestFixture.createReaction(feed, user, 2L)

            When("리액션 수를 3으로 업데이트하면") {
                feedUpdateService.updateTotalReaction(feed, reaction, user, 3L)

                Then("피드, 리액션, 사용자의 리액션 수가 증가해야 한다") {
                    feed.totalReactionCount shouldBe 3L
                    reaction.reactionCount shouldBe 5L
                    user.totalReactionCount shouldBe 3L
                }
            }
        }
    })
